_posts/2022-10-01-arrow-parquet-encoding-part-3.md

```diff
@@ -0,0 +1,169 @@
+limitations under the License.
+{% endcomment %}
+-->
+
+## Introduction
+
+This is the third of a three part series exploring how projects such as [Rust Apache Arrow](https://github.com/apache/arrow-rs) support conversion between [Apache Arrow](https://arrow.apache.org/) for in memory processing and [Apache Parquet](https://parquet.apache.org/) for efficient storage. This post covers how to combine the `Struct` and `List` types described in the previous posts for arbitrary nesting.
+
+
+[Apache Arrow](https://arrow.apache.org/) is an open, language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations. [Apache Parquet](https://parquet.apache.org/) is an open, column-oriented data file format designed for very efficient data encoding and retrieval.
+
+
+# Structs with Lists
+
+
+```json
+{                     <-- First record
+  “a”: [1],           <-- top-level field a containing list of integers
+  “b”: [              <-- top-level field b containing list of structures
+    {                 <-- list element of b containing two field b1 and b2
+      “b1”: 1         <-- b1 is always provided (not null)
+    },
+    {
+      “b1”: 1,
+      “b2”: [         <-- b2 contains list of integers
+        3, 4          <-- list elements of b.b2 always provided (not null)
+      ]
+    }
+  ]
+}
+{
+  “b”: [              <-- b is always provided (not null)
+    {
+      “b1”: 2
+    },
+  ]
+}
+{
+  “a”: [null, null],  <-- list elements of a are nullable
+  “b”: [null]         <-- list elements of b are nullable
+}
+```
+
+Documents of this format could be stored in this arrow schema
+
+```text
+Field(name: “a”, nullable: true, datatype: List(
+  Field(name: “element”, nullable: true, datatype: Int32),
+)
+Field(name: “b”), nullable: false, datatype: List(
+  Field(name: “element”, nullable: true, datatype: Struct[
+    Field(name: “b2”, nullable: false, datatype: Int32),
+    Field(name: “c2”, nullable: true, datatype: List(
+      Field(name: “element”, nullable: false, datatype: Int32)
+    ))
+  ])
+))
+```
+
+Documents of this format could be stored in this parquet schema
+
+```text
+message schema {
+  optional group a (LIST) {
+    repeated group list {
+      optional int32 element;
+    }
+  }
+  required group b (LIST) {
+    repeated group list {
+      optional group element {
+        required int32 c1;
+        optional group c2 (LIST) {
+          repeated group list {
+            required int32 element;
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+As explained previously, Arrow chooses to represent this in a hierarchical fashion. To achieve this it stores a list of monotonically increasing integers called offsets in the parent ListArray, and stores all the values that appear in the lists in a single child array. Each consecutive pair of elements in this offset array identifies a slice of the child array for that array index.
+
+```text
```

Comment:
I think making this into a diagram would be good -- I just haven't had a chance

Reply:
I don't know if it's helpful, but I made these diagrams when I was trying to build a mental image of `ArrayData` buffers (in C++):

[lucidchart diagram](https://lucid.app/lucidchart/b509b478-3206-439d-b252-3ac1426506b0/edit?viewport_loc=-117%2C-225%2C3449%2C1777%2C0_0&invitationId=inv_68f57577-d909-41df-a620-6b63980c6cbe)

feel free to adapt it for any other visualization

Reply:
Thanks -- those are some great diagrams (I think the color especially helps to call out certain features)

I think for now I am going to stick with the diagrams I have as I have already spent way too much time on this post 😆